src/frontend/js/components/SolutionData.js

```diff
@@ -1,48 +1,245 @@
 import React from 'react'
 import { graphql, createFragmentContainer } from 'react-relay'
+import { Button, Chip } from 'material-ui'
+import { withStyles } from 'material-ui/styles'
 import renderSolution from '../helpers/vexFlowSolutionRendering'
 import { constructThemeSolutionData } from '../helpers/constructThemeSolutionData'
+import DeletionModal from './DeletionModal'
+import updateSolutionMutation from './updateSolutionMutation'
+import deleteSolutionMutation from './deleteSolutionMutation'
+
+const styles = theme => ({
+  icon: {
+    width: 35,
+    height: 35,
+    marginLeft: 5,
+    fontSize: 30,
+    backgroundColor: '#BEEBEF',
+    '&:hover': {
+      backgroundColor: '#BEEBEF'
+    },
+    '&:focus': {
+      backgroundColor: '#BEEBEF'
+    }
+  },
+  image: {
+    height: 21,
+    width: 23,
+    marginBottom: 0,
+    paddingBottom: 0
+  },
+  whole: {
+    height: 21,
+    width: 26,
+    marginBottom: 0,
+    padding: 0,
+    paddingLeft: 4
+  },
+  button: {
+    marginLeft: 10
+  }
+})
 
 class SolutionData extends React.Component {
   constructor () {
     super()
     this.state = {
       cursor: {
         stave: 0,
-        voice: 0,
+        voice: '',
         position: 0
-      }
+      },
+      themeData: '',
+      solutionData: '',
+      saveButton: 'save',
+      openModal: false
     }
     this.eventListener = this.eventListener.bind(this)
+    this.handleModification = this.handleModification.bind(this)
+    this.modifyDuration = this.modifyDuration.bind(this)
+    this.openModal = this.openModal.bind(this)
+    this.closeModal = this.closeModal.bind(this)
+    this.deleteSolution = this.deleteSolution.bind(this)
+  }
+
+  saveSolution () {
+    updateSolutionMutation.commit(this.props, this.state.solutionData)
+    this.setState({saveButton: 'Saved'})
+    setTimeout(() => { this.setState({saveButton: 'Save'}) }, 2000)
+  }
+
+  deleteSolution () {
+    deleteSolutionMutation.commit(this.props)
+  }
+
+  openModal () {
+    this.setState({openModal: true})
+  }
+
+  closeModal () {
+    this.setState({openModal: false})
   }
 
   _renderSolution (element, solutionInfo) {
-    return renderSolution(element, solutionInfo, this.props.editable, this.state.cursor)
+    return renderSolution(element, solutionInfo, this.props.editable, this.state.cursor, this.props.givenVoice)
+  }
+
+  getNoteOctave () {
+    if (this.state.cursor.stave !== 0) {
+      return '/3'
+    } else {
+      return '/4'
+    }
   }
 
   eventListener (event) {
     const cursor = {...this.state.cursor}
-    switch (event.keyCode) {
-      case 39:
-        cursor.position += 1
-        this.setState({cursor: cursor})
-        break
-      case 37:
-        cursor.position = Math.max(cursor.position - 1, 0)
-        this.setState({cursor: cursor})
+    const voiceMapper = {
+      'soprano': 0,
+      'alto': 1,
+      'tenoro': 2
     }
+    const reverseVoiceMapper = {
```

Comment:
I'd be nice if this was calculated from `voiceMapper` for example using the `lodash`module's invert function https://lodash.com/docs/4.17.10#invert